SAI-4837 : In-memory synthetic `SolrCore` for QA node

patsonluk opened the pull request:
## Description
A prototype to attempt using an in-memory `SyntheticSolrCore`  similar to the `SolrCoreProxy` in our [Solr 8.8](https://github.com/fullstorydev/lucene-solr/blob/release/8.8/solr/core/src/java/org/apache/solr/core/SolrCoreProxy.java)

The synthetic collection/replica introduced in Solr 9 added the collection folder to physically disk and also the zookeeper `solr/collections`. 

However, it seems that what we essentially need is a `SolrCore` in the memory for each config set, which might be able to simply register it against the `CoreContainer` without actually registering to ZK.

## Solution
Introduced `SyntheticSolrCore` which extends `SolrCore`, `SyntheticSolrCore` is only used in Coordinator node to support a subset of `SolrCore` functionalities required by Coordinator operations such as aggregating and writing out response and providing configset info. 

Such core is only registered to the `CoreContainer` but is no longer register itself to zookeeper. 

Take note that to minimize code changes, there will still be snapshot/index data directory written to the QA node. We could have overridden `initSnapshotMetaDataManager` and `initIndex` to be no-op to avoid creating them, however, there is currently other code (and possibly future code too) that might rely on the existence of those values. It is safer to just keep them as is. 

The data folder for the synthetic core would NOT be loaded on QA node start-up as there's no `core.properties` in such directory. This is because instead of going through `CoreContainer#create` for such synthetic core, it's using `SyntheticSolrCore#createAndRegisterCore`, which bypass `core.properties` creation and zk registration


## Tests
1. `TestCoordinatorRole` passed, which provides good integration test coverage to ensure `SyntheticSolrCore` is working as expected
2. Local and playpen testing:
   i.  Basic query and restart tests from QA node
   ii. Shard split and moves and ensure QA node can distribute the requests to new shards
   iii. collection name is logged correctly
   iv. test on multiple collections to ensure that query work for them all, and the synthetic core should only be created once (by observing the log message `Loading synthetic core for config set...`)

patsonluk committed:
Experimental QA core as proxy in memory

patsonluk committed:
Experimental QA core as proxy in memory

patsonluk committed:
Experimental QA core as proxy in memory

patsonluk committed:
./gradlew tidy

patsonluk committed:
Comment out portion that checks for synthetic collection/core existence

hiteshk25 commented on solr/core/src/java/org/apache/solr/core/SolrCore.java:
```diff
@@ -1175,7 +1178,7 @@ private SolrCore(
 
       // Initialize the RestManager
-      restManager = initRestManager();
+      restManager = isSynthetic() ? new RestManager() : initRestManager();
```
how this is helping here?

patsonluk replied:
It was a quick prototype so I basically quickly handle NPEs when i ran into them. 😁 I put this one is so it will init w/o throwing exceptions.

I believe for a better implementation, we will need to understand all the init happened in the `SolrCore`, there's a fair chance that we can simply bypass most of the init routine for a synthetic core (rest manager, event handler/listener, plugin, searcher etc), as synthetic core does not do any real searching, it simply relays it to the actual data node

hiteshk25 commented on solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java:
```diff
@@ -66,6 +66,12 @@ public ZkConfigSetService(SolrZkClient zkClient) {
   }
 
+  @Override
+  protected SolrResourceLoader createConfigSetResourceLoader(CoreDescriptor cd, String configSetName) {
```
to understand better, why we need to override here?

patsonluk replied:
For synthetic core, the resource loader only needs to load the configset resource, it would have no access to collection specific resource - which is non existent for this new synthetic core/collection, since it does not even exists in ZK.

There are several implementation of `ConfigSetService`, and for prototype purpose, we only support this for `ZkConfigSetSrvice`

hiteshk25 commented on solr/core/src/java/org/apache/solr/servlet/CoordinatorHttpSolrCall.java:
```diff
@@ -85,118 +89,33 @@ public static SolrCore getCore(
-                e);
-          }
-        }
+        coreContainer.registerCore(syntheticCoreDescriptor, syntheticCore, false, false);
```
not publishing into zk? nice!

patsonluk committed:
Fixes watches and test cases

patsonluk commented:
Also pushed some fixes, so the unit test case would pass

patsonluk committed:
Refactoring to avoid modifying existing method modifiers

patsonluk committed:
Cleanup

patsonluk commented:
@chatman @noblepaul . I would ❤️ to gather some thoughts from you both! Mostly on the idea, as for the implementation details, we might need to further cleanup the logic if we decide to move forward.

The synthetic collection was created to avoid excess proxy cores created in 8.x AFAIK, I'm wondering if this proposal could address such concern without raising new ones?

noblepaul commented:
The biggest challenge is pushing this upstream. Making changes to `SolrCore` and introducing a concept of `synthetic` for a one of the core classes, for a niche feature is hardsell. However if this is to be only kept in-house , this should be fine

patsonluk committed:
Some refactoring to minimize changes to SolrCore

patsonluk committed:
javadoc

patsonluk commented:
> The biggest challenge is pushing this upstream. Making changes to `SolrCore` and introducing a concept of `synthetic` for a one of the core classes, for a niche feature is hardsell. However if this is to be only kept in-house , this should be fine

That's a good point, totally agree! Even if we do not upstream this, minimizing changes to `SolrCore` could save us from merge issues.

 I have added some minor changes and javadoc to my implementation. The change to `SolrCore` should now only be the access modifier of [`bufferUpdatesIfConstructing`](https://github.com/cowpaths/fullstory-solr/pull/181/files#diff-3bf44f923643744a743460bb0f64301618dd7529a602d945fe2f7f193d4cdde0). I'm hoping that the apache community will find it not as invasive (and it used to be `protected` as well in [8.8](https://github.com/fullstorydev/lucene-solr/blob/release/8.8/solr/core/src/java/org/apache/solr/core/SolrCore.java#L1135)).

patsonluk committed:
./gradlew tidy

patsonluk committed:
code cleanup

patsonluk committed:
code cleanup

patsonluk committed:
Fix resource loader issue if configSetName is null

patsonluk committed:
./gradlew tidy

noblepaul commented:
I like the fact that the changes to `SolrCore` is trivial/non-invasive. We can push this upstream as well

👍

noblepaul reviewed (approved):
Looks good

hiteshk25 commented on solr/core/src/java/org/apache/solr/core/SyntheticSolrCore.java:
```diff
@@ -0,0 +1,61 @@
+    // core's collection from ZK
+    // which synthetic core is not registered in ZK.
+    // We do not expect RestManager ops on Coordinator Nodes
+    return new RestManager();
```
what restManager enables?

noblepaul replied:
Actually, there is only a tiny feature called "managed resources" being powered by RestManager. It's not even reuired

hiteshk25 reviewed (approved):
LGTM 

(Thanks @patsonluk , please make sure QA logs are ok (collname, etc), though I don't feel any change in that area)

patsonluk committed:
Fixed issue with incorrect collection name in context
Ensure only one synthetic core created per config set

patsonluk committed:
./gradlew tidy

patsonluk commented:
@hiteshk25 

Good call on `please make sure QA logs are ok (collname, etc)` . The existing proposal did miss the collection name!

I have committed some fixes:
1. Collection name should be set properly for the collection being queried on now.
2. The `SyntheticSolrCore` should now be one instance per configset, my previous code didn't check correctly

Would you mind to review the changes in https://github.com/cowpaths/fullstory-solr/pull/181/commits/a32a2acb6a1524a061e40f3c2de53a4e6049e0ee . Many thanks again!!!

hiteshk25 reviewed (approved):
LGTM (Thanks @patsonluk )

patsonluk committed:
Correctly use <config set>_core as synthetic core name for clarity
Should open the registered core on first synthetic core creation since it is a getCore op

patsonluk committed:
./gradlew tidy

patsonluk committed:
Fixed issue with incorrect collection name in context
